### PR TITLE
Try fixing c++ compiler warnings about unknown union size

### LINKED ISSUE #13172: C++ compiler warning about size of `_jl_datatype_t`
I am seeing this warning when building Julia (OS X, clang 3.6.2):
```
./julia.h:289:5: warning: union has size 0 in C, non-zero size in C++
      [-Wextern-c-compat]
    union {
```
I see that this datatype is used from both C and C++ code.

One way to avoid this would be to add a `char dummy` to the union (and to update all routines that allocate memory for this type).

### PR DESCRIPTION
**Pull request by yuyichao**
I have a hard time reproducing the compiler warnings locally so I'm not 100% sure if this is the right fix....

Ref https://github.com/JuliaLang/julia/pull/13147#discussion_r39684851
Fix #13172

@vtjnash

### CODE REVIEW

**Comment by yuyichao** [comment]
Sadly the codegen rewrite didn't get rid of the old win32 llvm `SmallVector` assertion....
https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.9279/job/bqqikp508nx6anoc

Should be good otherwise?

**Comment by yuyichao** [comment]
@eschnett Can you verify if this fixes the compiler warning for you?

**Comment by eschnett** [comment]
I confirm that a test with this patch on top of release-0.4 did not lead to this warning.

**Comment by yuyichao** [comment]
Will merge later today if there isn't any other problems.

File: src/julia.h [hunk 1]
```diff
@@ -290,11 +290,7 @@ typedef struct _jl_datatype_t {
     uint32_t uid;
     void *struct_decl;  //llvm::Value*
     void *ditype; // llvm::MDNode* to be used as llvm::DIType(ditype)
-    union {
-        jl_fielddesc8_t fields8[0];
-        jl_fielddesc16_t fields16[0];
-        jl_fielddesc32_t fields32[0];
-    };
+    char fields[];
```
Comment (vtjnash) [inline src/julia.h]: `char` is going to have pretty pessimistic alignment assumptions. size_t might be a better choice.

Comment (yuyichao) [inline src/julia.h]: Done.

**Comment by yuyichao** [comment]
Any further comments?

**Comment by tkelman** [comment]
backported in bdf3cc4f81909a12b0551fa01694b4a950846d33

**Pull request merged**